DatadogInternal/Sources/NetworkInstrumentation/NetworkInstrumentationFeature.swift

```diff
@@ -0,0 +1,143 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+/// The Network Instrumentation Feature that can be registered into a core if
+/// any interceptor is provided.
+///
+/// Usage:
+///
+///     let core: DatadogCoreProtocol
+///
+///     let interceptor: DatadogURLSessionInterceptor = CustomDatadogTraceInterceptor()
+///     core.register(urlSessionInterceptor: interceptor)
+///
+/// Registering multiple interceptor will aggregate instrumentation.
+internal final class NetworkInstrumentationFeature: DatadogFeature {
+    /// The Feature name: "trace-propagation".
+    static let name = "network-instrumentation"
+
+    /// A no-op message bus receiver.
+    internal let messageReceiver: FeatureMessageReceiver = NOPFeatureMessageReceiver()
+
+    /// The list of registered propagators.
+    internal var interceptors: [DatadogURLSessionInterceptor] = []
+
+    /// Maps `URLSessionTask` to its `TaskInterception` object.
+    internal var interceptions: [URLSessionTask: URLSessionTaskInterception] = [:]
+
+    init() throws {
+        try URLSessionSwizzler.bind()
+    }
+
+    deinit {
+        URLSessionSwizzler.unbind()
+    }
+}
+
+extension NetworkInstrumentationFeature {
+    /// Tells the interceptors to modify a URL request.
+    ///
+    /// - Parameters:
+    ///   - session: The session [processing the request.
+    ///   - request: The request to intercept.
+    /// - Returns: The modified request.
+    func urlSession(_ session: URLSession, intercept request: URLRequest) -> URLRequest {
+        let headerTypes = firstPartyHosts(for: session).tracingHeaderTypes(for: request.url)
+        return interceptors.reduce(request) {
+            $1.modify(request: $0, headerTypes: headerTypes)
+        }
+    }
+
+    /// Tells the interceptors that the session did create a task.
+    ///
+    /// - Parameters:
+    ///   - session: The session creating the task.
+    ///   - task: The created task.
+    func urlSession(_ session: URLSession, didCreateTask task: URLSessionTask) {
+        guard let request = task.originalRequest else {
+            return
+        }
+
+        /// if any interceptor reject the request, then we stop here.
+        for interceptor in interceptors {
+            if interceptor.isInternal(request: request) {
+                return
+            }
+        }
+
+        let interception = URLSessionTaskInterception(
+            request: request,
+            isFirstParty: firstPartyHosts(for: session)
+                .isFirstParty(url: request.url)
+        )
+
+        interceptions[task] = interception
```

Comment:
Here we alter shared state from user-initiated `URLSession.dataTask(...)`. This can be called on any (user) thread, so it will crash 🔥 if not synchronising `interceptions[:]`.

Such behaviours are typically asserted in V1. Migrating this test should make it covered:
https://github.com/DataDog/dd-sdk-ios/blob/bdd0cef7e50b2fc478583365781f357cad963c17/Tests/DatadogTests/Datadog/URLSessionAutoInstrumentation/Interception/URLSessionInterceptorTests.swift#L515

Reply:
Of course, my bad!